internal/datasetaccess/checks.go

```diff
@@ -0,0 +1,150 @@
+package datasetaccess
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/gin-contrib/sessions"
+	"github.com/gin-gonic/gin"
+	"gopkg.in/yaml.v2"
+)
+
+const accessControlFilename = "openem-access.yaml"
+
+func parseAccessFile(path string) (accessFile, error) {
+	rawAccessFile, err := os.ReadFile(path)
+	if err != nil {
+		return accessFile{}, err
+	}
+
+	var parsedAccessFile accessFile
+	err = yaml.Unmarshal(rawAccessFile, &parsedAccessFile)
+	return parsedAccessFile, err
+}
+
+func CheckUserAccess(ctx context.Context, path string) error {
+	ginCtx, ok := ctx.(*gin.Context)
+	if !ok {
+		return fmt.Errorf("internal context error")
+	}
+
+	userSession := sessions.DefaultMany(ginCtx, "user")
+	userGroups, ok := userSession.Get("access_groups").([]string)
+	if !ok {
+		return fmt.Errorf("internal user session error: can't get access groups of user")
+	}
+
+	splitPath := strings.Split(filepath.Clean(path), string(filepath.Separator))
+	allowedGroups := map[string]bool{}
+	blockedGroups := map[string]bool{}
+	conflictedGroups := []string{}
+	invalidAllowGroups := []string{}
+
+	// assemble allowed and blocked group sets
+	for i := len(splitPath); i > 0; i-- {
+		/// read and parse file
+		currPath := filepath.Join(append(splitPath[0:i], accessControlFilename)...)
+		if splitPath[0] == "" {
+			currPath = string(filepath.Separator) + currPath // UNIX fix
+		}
+
+		if _, err := os.Stat(currPath); errors.Is(err, os.ErrNotExist) {
+			continue // skip the path if it doesn't contain any access restrictions
+		}
+
+		parsedAccessFile, err := parseAccessFile(currPath)
+		if err != nil {
+			return err
+		}
+
+		// path rule check
+		if parsedAccessFile.Path != currPath {
```

Comment:
I don't understand the use case for including `Path` at all if it is required to match the parent directory.